dask_match/core.py

```diff
@@ -616,7 +625,245 @@ def optimize_expr(expr):
             expr = replace_all(expr, replacement_rules)
     finally:
         _defer_to_matchpy = False
+
+    if fuse:
+        # Apply fusion
+        expr = _blockwise_fusion(expr)
+
     return expr
 
 
 from dask_match.reductions import Count, Max, Min, Mode, Size, Sum
+
+
+## Utilites for Expr fusion
+
+
+def replace_nodes(expr, nodes_to_replace: dict):
+    """Replace specific nodes in an Expr tree
+
+    `nodes_to_replace` corresponds to a `dict` mapping
+    of old `Expr` objects to new `Expr` objects.
+    """
+    if not nodes_to_replace:
+        # Nothing to replace
+        return expr
+    elif expr in nodes_to_replace:
+        # Already at targetted expr
+        return nodes_to_replace[expr]
+
+    new_operands = []
+    new_count = 0
+    for operand in expr.operands:
+        if isinstance(operand, Expr) and operand in nodes_to_replace:
+            # Replacing this operand
+            val = nodes_to_replace[operand]
+            new_count += 1
+        elif isinstance(operand, Expr):
+            # Non-Expr operand - Recursive call
+            val = replace_nodes(operand, nodes_to_replace)
+            if operand._name != val._name:
+                new_count += 1
+        else:
+            # Non-Expr operand
+            val = operand
+        new_operands.append(val)
+
+    if new_count:
+        # Only return new object if something changed
+        return type(expr)(*new_operands)
+    return expr
+
+
+def _blockwise_fusion(expr):
+    """Traverse the expression graph and apply fusion"""
```

Comment:
Is this doing diamond-y things?

Reply:
Yes (or at least that is my intention). I'd like to clean things up a bit, but the basic pseudo-code is:

- Start with an un-optimized `expr`
- (**Outer loop**) While not done:
  - Find the "dependents" of every `Fusable` node in `expr` (`dependents`)
  - Use `dependents` to construct a list of possible fusion roots (`roots`). Note that a "root" is a `Fusable` node in `expr` that has no `Fusable` dependents.
  - (**Inner loop**) For each `root` in `roots`:
    - Pop `root` from `roots` list 
    - Starting at `root`, traverse the tree of `Fusable` dependencies until we come to a node that is not eligible for fusion into the local group. We only allow a node to be fused into the group if it has no dependents outside of the known group. Since we traverse the graph in a breadth-first manner, we should be accounting for typical "diamond" dependency patterns.
    - Once we have found a node that cannot be fused into the current group, we have finished constructing the current `group`. If the ineligible node has remaining `Fusable` dependencies, we append it to the list of possible `roots`.
    - If `group` only contains `root`, we ignore the group and continue the loop at the next `root`
    - If `group` contains multiple nodes, we replace the `root` node of `expr` with a `FusedExpr` object (wrapping the original group)
    - If `roots` is empty, we are done fusing, and can break from the inner **and** outer loop
    - If `roots` is not empty, we need to run through the outer loop again, and only break from the **inner** loop
- Return `expr`, which is now fused.